source-hubspot-native/source_hubspot_native/api.py

```diff
@@ -216,137 +216,404 @@ async def fetch_batch_with_associations(
+'''
+
+async def fetch_changes_with_associations(
+    object_name: str,
+    cls: type[CRMObject],
+    fetcher: _FetchIdsFn,
+    log: Logger,
+    http: HTTPSession,
+    since: datetime,
+    until: datetime | None = None
+) -> AsyncGenerator[tuple[datetime, str, CRMObject], None]:
+    
     # Walk pages of recent IDs until we see one which is as-old
-    # as `log_cursor`, or no pages remain.
+    # as `since`, or no pages remain.
     recent: list[tuple[datetime, str]] = []
     next_page: PageCursor = None
-    using_fallback = False
 
     while True:
-        iter, next_page = await fetch_recent(log, http, log_cursor, next_page)
+        iter, next_page = await fetcher(next_page, len(recent))
 
         for ts, id in iter:
-            if ts > log_cursor:
+            if until and ts > until:
+                continue
+            elif ts > since:
+                # TODO(whb): It may be worth consulting the emitted changes
+                # cache here to see if we have already emitted a more recent
+                # change event before we do all the associations fetching work.
+                # Before implementing that I'd like to make sure that the
+                # top-level filtering works in production. Since the delayed
+                # changes stream only runs every 5 minutes or so it shouldn't be
+                # a huge load on the connector.
                 recent.append((ts, id))
             else:
                 next_page = None
 
         if not next_page:
             break
 
-        if fallback_to_search_api and not using_fallback and len(recent) >= 9_900:
-            log.info(f"falling back to Search API to return more than 10,000 recent records")
-
-            # Some fetch_recent functions are limited to a maximum number of
-            # recent records. After that point we can use the Search API to fill
-            # in the rest. The Search API is slower and has consistency problems
-            # with very recent data so it is generally not used as the first
-            # option for capturing data, but falling back to it is preferable
-            # over doing an entire re-backfill.
-            using_fallback = True
-            next_page = None
-
-            # The datetime of the oldest of the retrieved records is used here
-            # as the inclusive upper limit for the Search API. There may be many
-            # records with the same "created at" timestamp and runs of these
-            # records may span the `use_search_api_limit` boundary. In this case
-            # we will capture some of the same records twice.
-            fetch_recent = functools.partial(fetch_recent_search_objects, object_name, min(r[0] for r in recent))
-
     recent.sort()  # Oldest updates first.
 
     for batch_it in itertools.batched(recent, 50):
         batch = list(batch_it)
 
+        # Enable lookup of datetimes for IDs from the result batch.
+        dts = {r[1]: r[0] for r in batch}
```

Comment:
I think this can be rewritten as below for better readability
```suggestion
        dts = {id: dt for dt, id in batch}
```